### PR TITLE
[FLINK-26089][table-planner] Introduce TablePipeline

### PR DESCRIPTION
**Pull request by slinkydeveloper**
This PR introduces a new API object called `TablePipeline` which encapsulates the description of an table pipeline, as a result of a compilation of an `INSERT INTO`. The implementation is simply a wrapper for `List<ModifyOperation>`. It also adds the method `StatementSet#add(TablePipeline)` to add a pipeline to a statement set.

In order to improve the overall design quality of our APIs, in the first commit I introduced 3 interfaces splitting the capabilities of our API components:

* `Compilable`: an object that can be compiled to a `CompiledPlan`
* `Explainable`: an object that can be explained
* `Executable`: an object that can be executed

This brings overall consistency to our APIs and improves documentation. This change is a refactoring and doesn't bring any new feature, except the fluent `printExplain` method.

The implementation of these capabilities is:

|                 | `Compilable`       | `Executable`       | `Explainable`      |
|-----------------|--------------------|--------------------|--------------------|
| `Table`         | FLINK-25843        | :heavy_check_mark: | :heavy_check_mark: |
| `StatementSet`  | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| `CompiledPlan`  | Won't fix          | FLINK-26131   | :heavy_check_mark: |
| `TablePipeline` | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |

## Verifying this change

Added a test to verify a table pipeline can be compiled to a plan. The rest of the changes is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TablePipeline.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+package org.apache.flink.table.api;
+
+import org.apache.flink.annotation.PublicEvolving;
+
+/**
+ * This interface describes a pipeline created from a single {@code INSERT INTO} statement, created
+ * either using Table APIs or SQL.
+ */
+@PublicEvolving
+public interface TablePipeline extends Explainable<TablePipeline>, Executable, Compilable {}
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TablePipeline.java:28]: to fill this interface with something useful. how about we provide an `ObjectIdentifier` for the sink table.

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/api/bridge/java/StreamStatementSet.java [hunk 2]
```diff
@@ -38,18 +40,26 @@
 @PublicEvolving
 public interface StreamStatementSet extends StatementSet {
 
+    /** {@inheritDoc} */
```
Comment (twalthr) [inline flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/api/bridge/java/StreamStatementSet.java:43]: This should actually not be necessary. I remember that I also did some research. Did you verify this with IntelliJ?

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/api/bridge/java/StreamStatementSet.java [hunk 3]
```diff
@@ -38,18 +40,26 @@
...
     @Override
     StreamStatementSet addInsertSql(String statement);
 
+    /** {@inheritDoc} */
     @Override
+    StreamStatementSet add(TablePipeline tablePipeline);
```
Comment (twalthr) [inline flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/api/bridge/java/StreamStatementSet.java:49]: nit: add it to the end of all `insert` methods or to the beginning

File: flink-table/flink-table-api-scala-bridge/src/main/scala/org/apache/flink/table/api/bridge/scala/internal/StreamStatementSetImpl.scala [hunk 4]
```diff
@@ -21,40 +21,39 @@ package org.apache.flink.table.api.bridge.scala.internal
...
 /** Implementation for [[StreamStatementSet]]. */
 @Internal
 class StreamStatementSetImpl(tableEnvironment: StreamTableEnvironmentImpl)
     extends StatementSetImpl[StreamTableEnvironmentImpl](tableEnvironment)
     with StreamStatementSet {
 
-  override def addInsertSql(statement: String): StreamStatementSet = {
+  override def addInsertSql(statement: String): StreamStatementSet =
     super.addInsertSql(statement).asInstanceOf[StreamStatementSet]
-  }
```
Comment (twalthr) [inline flink-table/flink-table-api-scala-bridge/src/main/scala/org/apache/flink/table/api/bridge/scala/internal/StreamStatementSetImpl.scala:34]: nit: IMHO we should leave curly braces for methods, we don't need to use very fancy Scala syntax, if you would add another statement, you would add the statement to the class init, pretty dangerous.

File: flink-table/flink-table-api-scala-bridge/src/main/scala/org/apache/flink/table/api/bridge/scala/internal/StreamStatementSetImpl.scala [hunk 5]
```diff
@@ -21,40 +21,39 @@ package org.apache.flink.table.api.bridge.scala.internal
...
+  override def addInsert(targetDescriptor: TableDescriptor, table: Table): StreamStatementSet =
     super.addInsert(targetDescriptor, table).asInstanceOf[StreamStatementSet]
-  }
 
   override def addInsert(
-      targetDescriptor: TableDescriptor,
-      table: Table,
-      overwrite: Boolean)
-    : StreamStatementSet = {
+                          targetDescriptor: TableDescriptor,
```
Comment (twalthr) [inline flink-table/flink-table-api-scala-bridge/src/main/scala/org/apache/flink/table/api/bridge/scala/internal/StreamStatementSetImpl.scala:48]: fix indention

File: flink-table/flink-table-api-scala-bridge/src/main/scala/org/apache/flink/table/api/bridge/scala/internal/StreamStatementSetImpl.scala [hunk 6]
```diff
@@ -21,40 +21,39 @@ package org.apache.flink.table.api.bridge.scala.internal
...
+                          table: Table,
+                          overwrite: Boolean)
+  : StreamStatementSet =
     super.addInsert(targetDescriptor, table, overwrite).asInstanceOf[StreamStatementSet]
+
+  override def printExplain(extraDetails: ExplainDetail*): StreamStatementSet = {
+    System.out.println(super.explain(extraDetails: _*))
```
Comment (twalthr) [inline flink-table/flink-table-api-scala-bridge/src/main/scala/org/apache/flink/table/api/bridge/scala/internal/StreamStatementSetImpl.scala:55]: Scala offers `println()` without `System.out`

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/api/CompiledPlanITCase.java [hunk 7]
```diff
@@ -95,7 +96,19 @@ public void testExecutePlan() throws Exception {
     }
 
     @Test
-    public void testExecutePlanSql() throws Exception {
+    public void testExecutePlanTable() throws Exception {
+        List<String> data = Arrays.asList("1,1,hi", "2,1,hello", "3,2,hello world");
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/api/CompiledPlanITCase.java:100]: use constants above

**Review by twalthr — CHANGES_REQUESTED** [review]
Thanks for the PR @slinkydeveloper. I added mostly minor feedback. Should be good in the next iteration.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TablePipeline.java (hunk 1 above)
Comment (slinkydeveloper) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TablePipeline.java:28]: I added both `getSinkIdentifier` and implemented `toString` in `TablePipelineImpl`

**Pull request closed without merging**